apiserver/facades/client/secrets/secrets.go

```diff
@@ -406,16 +407,48 @@ func (s *SecretsAPI) updateSecret(backend provider.SecretsBackend, arg params.Up
 			}
 		}
 	}
-	_, err = s.secretsState.UpdateSecret(uri, fromUpsertParams(arg.UpsertSecretArg))
-	return errors.Trace(err)
+	md, err = s.secretsState.UpdateSecret(uri, fromUpsertParams(arg.AutoPrune, arg.UpsertSecretArg))
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if md.AutoPrune {
+		// If the secret was updated, we need to delete the old unused secret revisions.
+		revsToDelete, err := s.secretsState.ListUnusedSecretRevisions(uri)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		pruneArg := params.DeleteSecretArg{URI: md.URI.String()}
+		for _, rev := range revsToDelete {
+			if rev == md.LatestRevision {
+				// We don't want to delete the latest revision.
+				continue
+			}
+			pruneArg.Revisions = append(pruneArg.Revisions, rev)
+		}
+		if len(pruneArg.Revisions) == 0 {
+			return nil
+		}
+		pruneResult, err := s.RemoveSecrets(params.DeleteSecretArgs{Args: []params.DeleteSecretArg{pruneArg}})
+		if err != nil {
+			// We don't want to fail the update if we can't prune the unused secret revisions because they will be picked up later
+			// when the secret has any new obsolute revisions.
+			logger.Warningf("failed to prune unused secret revisions for %q: %v", uri, err)
+		}
+		if err = pruneResult.Combine(); err != nil {
+			logger.Warningf("failed to prune unused secret revisions for %q: %v", uri, pruneResult.Combine())
+		}
```

Comment:
See above - we should surface this to the caller. They asked for prune and it failed so they need to know.